src/join.sol

```diff
@@ -427,3 +427,93 @@ contract AuthGemJoin is LibNote {
         require(gem.transfer(usr, wad), "AuthGemJoin/failed-transfer");
     }
 }
+
+interface GemLike7 {
+    function decimals() external view returns (uint);
+    function transfer(address,uint) external;
+    function transferFrom(address,address,uint) external;
+    function balanceOf(address) external view returns (uint);
+    function allowance(address,address) external view returns (uint);
+    function upgradedAddress() external view returns (address);
+    function deprecated() external view returns (bool);
+    function setImplementation(address,uint) external;
+}
+
+contract GemJoin7 is LibNote {
+    mapping (address => uint) public wards;
+    function rely(address usr) external note auth { wards[usr] = 1; }
+    function deny(address usr) external note auth { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    VatLike  public vat;
+    bytes32  public ilk;
+    GemLike7 public gem;
+    uint     public dec;
+    uint     public live; // Access flag
+
+    mapping (address => uint256) public implementations;
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        gem = GemLike7(gem_);
+        dec = gem.decimals();
+        require(dec < 18, "GemJoin7/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        setImplementation(address(gem.upgradedAddress()), 1);
+    }
+
+    function cage() external note auth {
+        live = 0;
+    }
+
+    function setImplementation(address implementation, uint256 permitted) public auth note {
+        implementations[implementation] = permitted; // 1 live, 0 distable
+    }
+
+    function mul(uint x, uint y) internal pure returns (uint z) {
+        require(y == 0 || (z = x * y) / y == x, "GemJoin7/overflow");
+    }
+
+    function join(address urn, uint wad) public note {
+        require(live == 1, "GemJoin7/not-live");
+        // mul does overflow check so require(wad < 2 ** 255) not needed
```

Comment:
Hmm not sure exactly what this comment means. The `< 2 ** 255` is for checking if a `uint` can be cast to `int`, which happens with the `int(wad18) >= 0`.
The `mul` itself checks a `uint` overflow but not an `int` one.

Reply:
the `require(int(wad18) >= 0, GemJoin7/overflow")` checks the int overflow, yes? and the `mul()` checks the `uint` overflow, so there are sufficient checks in place?

i suppose my comment was more me thinking thru the logic